pyvista/core/utilities/points.py

```diff
@@ -459,3 +464,91 @@ def vector_poly_data(orig, vec):
+    or the right singular vectors from the Singular Value Decomposition (SVD).
+
+    The axes are computed as the eigenvectors of the covariance matrix from the
+    mean-centered points, and are processed to ensure that they form a right-handed
+    coordinate frame.
+
+    The axes explain the total variance of the points. The first axis explains the
+    largest percentage of variance, followed by the second axis, followed again by
+    the third axis which explains the smallest percentage of variance. The axes may
+    be used to build an oriented bounding box or to align the points to another set
+    of axes (e.g. the world XYZ axes).
+
+    .. note::
+        The computed axes are not unique, and the sign of each axis direction can be
+        arbitrarily changed.
+
+    .. versionadded:: 0.45.0
+
+    Parameters
+    ----------
+    points : MatrixLike[float]
+        Nx3 array of points.
+
+    return_sizes : bool, default: False
+        If ``True``, also returns the axes sizes. The sizes are computed as the square
+        root of the eigen vectors of the mean-centered covariance matrix.
+
+    Returns
+    -------
+    numpy.ndarray
+        3x3 orthonormal array with the principal axes as row vectors.
+
+    numpy.ndarray
+        Three-item array of the axes sizes.
+
+    Examples
+    --------
+    Create a mesh with points that have the largest variation in ``X``,
+    followed by ``Y``, then ``Z``.
+
+    >>> import pyvista as pv
+    >>> mesh = pv.ParametricEllipsoid(xradius=10, yradius=5, zradius=1)
+    >>> p = pv.Plotter()
+    >>> _ = p.add_mesh(mesh)
+    >>> _ = p.show_grid()
+    >>> p.show()
+
+    Compute its principal axes.
+
+    >>> principal_axes = pv.principal_axes(mesh.points)
+    >>> principal_axes
+    array([[-1.0000000e+00,  1.6606515e-10,  1.9055901e-10],
+           [-1.6606515e-10, -1.0000000e+00, -5.0257620e-10],
+           [ 1.9055901e-10, -5.0257620e-10,  1.0000000e+00]], dtype=float32)
+
+    Note that the principal axes have ones along the diagonal and zeros
+    in the off diagonals. This indicates that the first principal axis is
+    aligned with the x-axis, the second with the y-axis, and third with
+    the z-axis. This is expected, since the mesh is already axis-aligned.
+
+    However, since the signs of the principal axes are arbitrary, the
+    first and third axes in this case have a negative direction.
+
+    """
+    points = _validation.validate_arrayNx3(points)
+
+    points_centered = points - np.mean(points, axis=0)
+    eig_vals, eig_vectors = np.linalg.eigh(points_centered.T @ points_centered)
+    sizes = np.sqrt(eig_vals)[::-1]  # ascending order -> descending order
+    axes = eig_vectors.T[::-1]  # columns, ascending order -> rows, descending order
+
+    # Normalize and ensure output forms right-handed coordinate frame
+    axes[0] /= np.linalg.norm(axes[0])
+    axes[1] /= np.linalg.norm(axes[1])
+    axes[2] = np.cross(axes[0], axes[1])
```

Comment:
According [to the docs](https://numpy.org/doc/stable/reference/generated/numpy.linalg.eigh.html#numpy.linalg.eigh) the vectors are orthonormal. With that in mind I think we can skip normalization. Unfortunately neither the NumPy docs [nor that of LAPACK](https://netlib.org/lapack/explore-html/d8/d30/group__heevd_ga3cac18a55c8b7b87a42cfa26c1f3499a.html) mentions handedness. But I'd assume it's always right-handed, and if we still want to verify this we could also compute `np.linalg.det(axes)`. If it's negative, we have a left-handed system and we should flip the sign of one of the vectors (last one?). In that case we don't touch the vectors in most expected (right-handed) cases, and when we do touch them we only introduce sign changes rather than new floating-point errors (not that I'd expect these to be significant in any way).

Reply:
> According [to the docs](https://numpy.org/doc/stable/reference/generated/numpy.linalg.eigh.html#numpy.linalg.eigh) the vectors are orthonormal. With that in mind I think we can skip normalization. 

Might have been left over from the vtk implementation. The unit vector tests were failing at one point but I'll remove the normalization. Tests are passing without it.

> Unfortunately neither the NumPy docs [nor that of LAPACK](https://netlib.org/lapack/explore-html/d8/d30/group__heevd_ga3cac18a55c8b7b87a42cfa26c1f3499a.html) mentions handedness. But I'd assume it's always right-handed, 

Several of the tests fail when I comment out the cross-product, so this assumption can't be made.

> and if we still want to verify this we could also compute `np.linalg.det(axes)`. If it's negative, we have a left-handed system and we should flip the sign of one of the vectors (last one?). In that case we don't touch the vectors in most expected (right-handed) cases, and when we do touch them we only introduce sign changes rather than new floating-point errors (not that I'd expect these to be significant in any way).

I'm fine with this. Related to an earlier note about the numerical stability of SVD vs PCA, it does seem like there are precision issues with this method, Not huge, but I did have to tweak the tests to use "nicer" values, otherwise in at least one case I had to use `atol=1e-2` to make it pass. This was mostly an issue with values close to zero, and for low-rank inputs, and the error was very machine dependent. But you will still see `atol=1e-7` in the tests for this reason.

With that in mind, re-computing the third vector may help reduce the error in some cases. But, I think it's ok without this, and only change the sign if needed.

Reply:
The handedness of the axes was the important thing, so let's leave the vectors alone, the determinant check and direction swap has been added.

No more normalization or cross products now.